File: Banco/src/gestorAplicacion/entidades_de_negocio/Cliente.java
//Autores:
//Jose Miguel Pulgarin A.
//Carlos Guarin
//Dario Alexander Penagos V.

package gestorAplicacion.entidades_de_negocio;

import java.util.Arrays;

import gestorAplicacion.infraestructura.*;
import gestorAplicacion.tarjetas.*;

import java.io.Serializable;
import java.util.ArrayList;

public class Cliente implements Serializable{
	public final String nombre;
	private int Id;
	private ArrayList<TarjetaDebito> tarjetasDebito = new ArrayList<>();
	private ArrayList<TarjetaCredito> tarjetasCredito = new ArrayList<>();
	private ArrayList<Factura> facturas = new ArrayList<>();
	private int bonoActual = 0;
	
	public Cliente(String nombre, int noDeIdentificacion) {
		this.nombre = nombre;
		this.Id = noDeIdentificacion;
		Banco.agregarCliente(this);
	}
	
	
	//Metodos de la instancia
	
	@Override
	public String toString(){
		return "\n\nNombre: %s\nIdentificacion: %s\nNúmero de Tarjetas de Debito: %s\nNúmero de Tarjetas de Credito: %s\nBono Actual: %s".formatted(nombre, Id, tarjetasDebito.size(), tarjetasCredito.size(), bonoActual);
	}

	//Getters & Setters
	public int getNoDeIdentificacion() {
		return Id;
	}

	public void setNoDeIdentificacion(int noDeIdentificacion) {
		this.Id = noDeIdentificacion;
	}

	public ArrayList<TarjetaDebito>  getTarjetasDebito() {
		return tarjetasDebito;
	}

	public void agregarTarjetDebito(TarjetaDebito tarjetaDebito) {
		this.tarjetasDebito.add(tarjetaDebito);
	}
	
	public void agregarTarjetasDebito(ArrayList<TarjetaDebito> tarjetasDebito) {//Agregar varias tarjetas de un arrayList
		this.tarjetasDebito.addAll(tarjetasDebito);
	}
	
	public void agregarTarjetasDebito(TarjetaDebito... tarjetasDebito) {//Agregar varios tarjetas una por una Ej. (tarjeta1, tarjeta2, etc)
		this.tarjetasDebito.addAll(Arrays.asList(tarjetasDebito));
	}
	
	public void eliminarTarjetaDebito(TarjetaDebito tarjetaDebito) {
		this.tarjetasDebito.remove(tarjetaDebito);
	}
	
	public  void eliminarTarjetaDebito(ArrayList<TarjetaDebito> tarjetasDebito) {//Eliminar varias tarjetas de un ArrayList
		for(TarjetaDebito tarjetaDebito: tarjetasDebito) {
			this.tarjetasDebito.remove(tarjetaDebito);
		}			
	}
	
	public void eliminarTarjetaDebito(TarjetaDebito[] tarjetasDebito) {//Eliminar varias tarjetas de un Array Ej (tarjeta1, tarjeta2, etc)
		for(TarjetaDebito tarjetaDebito: tarjetasDebito) {
			this.tarjetasDebito.remove(tarjetaDebito);
		}			
	}

	public ArrayList<TarjetaCredito> getTarjetasCredito() {
		return tarjetasCredito;
	}

	public ArrayList<Tarjeta> getTarjetas() {
		ArrayList<Tarjeta> retorno = new ArrayList<>();
		retorno.addAll(tarjetasDebito);
		retorno.addAll(tarjetasCredito);
		return retorno;
	}

	public void agregarTarjetasCredito(TarjetaCredito tarjetaCredito) {
		this.tarjetasCredito.add(tarjetaCredito);
	}
	
	public void agregarTarjetasCredito(ArrayList<TarjetaCredito> tarjetasCredito) {//Agregar varias tarjetas de un arrayList
		this.tarjetasCredito.addAll(tarjetasCredito);
	}
	
	public void agregarTarjetasCredito(TarjetaCredito... tarjetasCredito) {//Agregar varios tarjetas una por una Ej. (tarjeta1, tarjeta2, etc)
		this.tarjetasCredito.addAll(Arrays.asList(tarjetasCredito));
	}
	
	public void eliminarTarjetaCredito(TarjetaCredito tarjetaCredito) {
		this.tarjetasCredito.remove(tarjetaCredito);
	}
	
	public  void eliminarTarjetasCredito(ArrayList<TarjetaCredito> tarjetasCredito) {//Eliminar varias tarjetas de un ArrayList
		for(TarjetaCredito tarjetaCredito: tarjetasCredito) {
			this.tarjetasCredito.remove(tarjetaCredito);
		}			
	}
	
	public void eliminarTarjetasCredito(TarjetaCredito[] tarjetasCredito) {//Eliminar varias tarjetas de un Array Ej (tarjeta1, tarjeta2, etc)
		for(TarjetaCredito tarjetaCredito: tarjetasCredito) {
			this.tarjetasCredito.remove(tarjetaCredito);
		}			
	}

	public ArrayList<Factura> getFactura() {
		return facturas;
	}

	public void agregarFactura(Factura factura) {
		this.facturas.add(factura);
	}
	
	public void eliminarFactura(Factura factura) {
		this.facturas.remove(factura);
	}

	public int getBonoActual() {
		return bonoActual;
	}

	public void setBonoActual(int bonoActual) {
		this.bonoActual = bonoActual;
	}
	
	//Metodos de las intancias
	
	public ArrayList<Factura> listarFacturas(){//Este metodo solo retorna facturas pendientes
		ArrayList<Factura> retorno = new ArrayList<>();
		for(Factura f : this.facturas){
			if(f.isFacturaVencida() && f.isFacturaPagada()){
				retorno.add(f);
			}
		}
		for(Factura f : this.facturas){
			if(!f.isFacturaVencida() && f.isFacturaPagada()){
				retorno.add(f);
			}
		}
		return retorno;
	}
	
	//Retorna las tarjetas de credito y debito que sean compatibles con la divisa de la factura
	//Ademas que no tengan fondos en 0 o creditoMaximo alcanzado
	//La tarjeta para ser listada también debe tener un estado "Activo"
	public ArrayList<Tarjeta> listarTarjetas(Factura factura){//Listar tarjetas para pagar factura
		ArrayList<Tarjeta> tarjetasDisponibles = new ArrayList<>();
		for(TarjetaDebito tarjeta : tarjetasDebito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(factura.getDIVISA())){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		for(TarjetaCredito tarjeta : tarjetasCredito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(factura.getDIVISA())){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		return tarjetasDisponibles;
	}
	
	public ArrayList<Tarjeta> listarTarjetas(ArrayList<Divisa> divisas){//Listar tarjetas para cambiar divisas. Recibe un arrayList
		Divisa divisaOrigen = divisas.get(0);
		Divisa divisaDestino = divisas.get(1);
		
		ArrayList<Tarjeta> tarjetasDisponibles = new ArrayList<>();
		
		for(TarjetaDebito tarjeta : tarjetasDebito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(divisaOrigen) || tarjeta.getDivisa().equals(divisaDestino)){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		
		for(TarjetaCredito tarjeta : tarjetasCredito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(divisaOrigen) || tarjeta.getDivisa().equals(divisaDestino)){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		
		return tarjetasDisponibles;
	}
	
	//El usuario debe tener tarjetas con la divisa de origen que escoja
	//Dichas tarjetas deben tener saldo o credito, y estar activas
	public ArrayList<Divisa> escogerDivisas(Divisa origen, Divisa destino) {//Retorna un ArrayList
		ArrayList<Divisa> divisas = new ArrayList<>();
		divisas.add(origen);
		divisas.add(destino);
		
		//Verificar que el cliente tenga una tarjeta de origen con la divisa que escogió
		for(TarjetaDebito tarjetaDebito: this.getTarjetasDebito()) {
			if(!tarjetaDebito.isActiva())
				continue;
			if(!tarjetaDebito.tieneSaldo())
				continue;
			if(!tarjetaDebito.getDivisa().equals(origen))
				continue;
			//Si existe al menos una tarjeta que cumpla con lo requerido, retornamos.
			return divisas;
		}
		
		//Si no existen aun tarjetas con dicha divisa (ya que no se retornó) se recorren las tarjetas de credito
		for(TarjetaCredito tarjetaCredito: this.getTarjetasCredito()) {
			if(!tarjetaCredito.isActiva())
				continue;
			if(!tarjetaCredito.tieneSaldo())
				continue;
			if(!tarjetaCredito.getDivisa().equals(origen))
				continue;
			return divisas;
		}
		
		//Si no se retornó anteriormente, es porque el cliente no tiene tarjetas con dicha divisa
		return null;
	}
	
	public Divisa[] escogerDivisas2(Divisa origen, Divisa destino) {//Retorna un Array normal
		Divisa[] divisas = new Divisa[2];
		divisas[0] = origen;
		divisas[1] = destino;
		
		//Verificar que el cliente tenga una tarjeta de origen con la divisa que escogió
		for(TarjetaDebito tarjetaDebito: this.getTarjetasDebito()) {
			if(!tarjetaDebito.isActiva())
				continue;
			if(!tarjetaDebito.tieneSaldo())
				continue;
			if(!tarjetaDebito.getDivisa().equals(origen))
				continue;
			//Si existe al menos una tarjeta que cumpla con lo requerido, retornamos.
			return divisas;
		}
		
		//Si no existen aun tarjetas con dicha divisa (ya que no se retornó) se recorren las tarjetas de credito
		for(TarjetaCredito tarjetaCredito: this.getTarjetasCredito()) {
			if(!tarjetaCredito.isActiva())
				continue;
			if(!tarjetaCredito.tieneSaldo())
				continue;
			if(!tarjetaCredito.getDivisa().equals(origen))
				continue;
			return divisas;
		}
		
		//Si no se retornó anteriormente, es porque el cliente no tiene tarjetas con dicha divisa
		return null;
	}
	
	public ArrayList<Transaccion> verPeticiones(){
		ArrayList<Transaccion> retorno = new ArrayList<>();
		for(Transaccion t : Transaccion.getTransacciones()){
			if(t.getClienteObjetivo() != null){		//Verifica si la transacción tiene un cliente objetivo. sin esta verificación, la siguiente línea podría soltar un error
				if(t.getClienteObjetivo().equals(this) && t.getMensaje() != null && t.isPendiente()){
					retorno.add(t);
				}
			}
			
		}
		return retorno;
	}

	public ArrayList<Canal> listarCanales(Divisa[] divisas) {//Recibe un array normal
		Divisa divisaOrigen = divisas[0];
		Divisa divisaDestino = divisas[1];
		
		ArrayList<Canal> canales = new ArrayList<>();
		for(Canal canal: Banco.getCanales()) {
			if(canal.tieneDivisa(divisaOrigen))
				continue;
			if(canal.tieneDivisa(divisaDestino))
				continue;
			if(canal.tieneFondosDeDivisa(divisaDestino))
				continue;
			canales.add(canal);
		}
		
		//ordenar por menos impuestos de menor a mayor
		return Banco.ordenarCanalesPorImpuestos(canales);
	}

	public ArrayList<Canal> listarCanales(ArrayList<Divisa> divisas) {//recibe un arrayList
		Divisa divisaOrigen = divisas.get(0);
		Divisa divisaDestino = divisas.get(1);
		
		ArrayList<Canal> canales = new ArrayList<>();
		for(Canal canal: Banco.getCanales()) {
			if(canal.tieneDivisa(divisaOrigen))
				continue;
			if(canal.tieneDivisa(divisaDestino))
				continue;
			if(canal.tieneFondosDeDivisa(divisaDestino))
				continue;
			canales.add(canal);
		}
		
		//ordenar por menos impuestos de menor a mayor
		return Banco.ordenarCanalesPorImpuestos(canales);
	}
	
	public ArrayList<Transaccion> revisarHistorialCreditos(){
		ArrayList<Transaccion> retorno = new ArrayList<>();
		for(Transaccion t : Transaccion.getTransacciones()){
			if(t.getClienteOrigen().equals(this) && t.getTarjetaOrigen() instanceof TarjetaCredito && !t.isRechazado() && !t.isPendiente()){
				retorno.add(t);
			}
		}
		return retorno;
	}
}

File: versionEnPython/Banco/src/gestorAplicacion/Entidades_de_negocio/Cliente.py
from Banco.src.gestorAplicacion.Entidades_de_negocio.Transaccion import Transaccion


class Cliente:
    def __init__(self, nombre, noDeIdentificacion):
        """
        Constructor de la clase Cliente.

        Args:
            nombre (str): Nombre del cliente.
            noDeIdentificacion (int): Número de identificación del cliente.
        """
        from Banco import Banco
        self.nombre = nombre
        self.Id = noDeIdentificacion
        self.tarjetasDebito = []
        self.tarjetasCredito = []
        self.facturas = []
        Banco.agregarCliente(self)

    def getNombre(self):
        """
        Devuelve el nombre del cliente.

        Returns:
            str: Nombre del cliente.
        """
        return self.nombre

    def revisarHistorialDeCreditos(self):
        """
        Revisa el historial de transacciones de crédito del cliente.

        Returns:
            list: Lista de transacciones de crédito realizadas por el cliente.
        """
        from Banco.src.gestorAplicacion.tarjetas.TarjetaCredito import TarjetaCredito
        retorno = []
        for t in Transaccion.getTransacciones():
            if t.getClienteOrigen() == self and not t.rechazado and isinstance(t.getTarjetaOrigen(),TarjetaCredito) and not t.pendiente:
                retorno.append(t)
        return retorno

    def getTarjetasCredito(self):
        """
        Devuelve las tarjetas de crédito del cliente.

        Returns:
            list: Lista de tarjetas de crédito del cliente.
        """
        return self.tarjetasCredito

    def getTarjetasDebito(self):
        """
        Devuelve las tarjetas de débito del cliente.

        Returns:
            list: Lista de tarjetas de débito del cliente.
        """
        return self.tarjetasDebito

    def getFacturas(self):
        """
        Devuelve las facturas del cliente.

        Returns:
            list: Lista de facturas del cliente.
        """
        return self.facturas

    def agregarFactura(self, f):
        """
        Agrega una factura a la lista de facturas del cliente.

        Args:
            f (Factura): Factura a agregar.
        """
        self.facturas.append(f)

    def getTarjetas(self):
        """
        Devuelve todas las tarjetas del cliente.

        Returns:
            list: Lista de todas las tarjetas del cliente.
        """
        return self.tarjetasCredito + self.tarjetasDebito

    def listarTarjetas(self, factura):
        """
        Lista las tarjetas disponibles para pagar una factura en una divisa determinada.

        Args:
            factura (Factura): Factura a pagar.

        Returns:
            list: Lista de tarjetas disponibles para pagar la factura.
        """
        tarjetasDisponibles = []
        for tarjeta in self.tarjetasDebito:
            if not tarjeta.tieneSaldo():
                continue
            if tarjeta.getDivisa() == factura.getDIVISA():
                tarjetasDisponibles.append(tarjeta)
        for tarjeta in self.tarjetasCredito:
            if not tarjeta.tieneSaldo():
                continue
            if tarjeta.getDivisa() == factura.getDIVISA():
                tarjetasDisponibles.append(tarjeta)
        return tarjetasDisponibles

    def agregarTarjetasDebito(self, *tarjetasDebito):
        """
        Agrega una o varias tarjetas de débito al cliente.

        Args:
            *tarjetasDebito: Tarjetas de débito a agregar.
        """
        self.tarjetasDebito.extend(list(tarjetasDebito))

    def agregarTarjetasCredito(self, *tarjetasCredito):
        """
        Agrega una o varias tarjetas de crédito al cliente.

        Args:
            *tarjetasCredito: Tarjetas de crédito a agregar.
        """
        self.tarjetasCredito.extend(list(tarjetasCredito))

    def encontrarTarjeta(self, tarjeta):
        """
        Encuentra una tarjeta basándose en su representación en cadena (toString).

        Args:
            tarjeta: Tarjeta a buscar.

        Returns:
            Tarjeta: Tarjeta encontrada o None si no se encuentra.
        """
        for t in self.tarjetasCredito:
            if t.__str__() == tarjeta.__str__():
                return t
        for t in self.tarjetasDebito:
            if t.__str__() == tarjeta.__str__():
                return t

    def encontrarFacturas(self, factura):
        """
        Encuentra una factura basándose en su representación en cadena (toString).

        Args:
            factura: Factura a buscar.

        Returns:
            Factura: Factura encontrada o None si no se encuentra.
        """
        for f in self.facturas:
            if f.__str__() == factura.__str__():
                return f

    def mostrarFacturas(self):
        """
        Muestra las facturas pendientes de pago del cliente.

        Returns:
            list: Lista de facturas pendientes de pago.
        """
        retorno = []
        for f in self.facturas:
            if not f.facturaPagada:
                retorno.append(f)
        return retorno

    def verPeticiones(self):
        """
        Muestra las transacciones pendientes del cliente que tienen un mensaje adjunto.

        Returns:
            list: Lista de transacciones pendientes con mensaje.
        """
        retorno = []
        for t in Transaccion.getTransacciones():
            if t.getClienteObjetivo() is not None:
                if t.getClienteObjetivo() == self and t.getMensaje() is not None and t.isPendiente():
                    retorno.append(t)
        return retorno

    def tarjetasConDivisa(self, divisa, origen: bool):
        """
        Devuelve una lista de tarjetas del cliente que tienen una divisa determinada.

        Args:
            divisa (str): Divisa a buscar.
            origen (bool): Indica si se deben incluir las tarjetas de crédito (True) o no (False).

        Returns:
            list: Lista de tarjetas con la divisa especificada.
        """
        retorno = []
        for t in self.tarjetasDebito:
            if not t.isActiva():
                continue
            if not t.tieneSaldo():
                continue
            if t.getDivisa() == divisa:
                retorno.append(t)
        if origen:
            for t in self.tarjetasCredito:
                if not t.isActiva():
                    continue
                if not t.tieneSaldo():
                    continue
                if t.getDivisa() == divisa:
                    retorno.append(t)
        return retorno

    def listarCanales(self, divisao, divisad):
        """
        Lista los canales disponibles para realizar una transacción entre dos divisas.

        Args:
            divisao (str): Divisa de origen.
            divisad (str): Divisa de destino.

        Returns:
            list: Lista de canales disponibles ordenados por impuestos.
        """
        from Banco import Banco
        divisaOrigen = divisao
        divisaDestino = divisad
        canales = []
        for canal in Banco.getCanales():
            if canal.tieneDivisa(divisaOrigen):
                continue
            if canal.tieneDivisa(divisaDestino):
                continue
            canales.append(canal)
        return Banco.ordenarCanalesPorImpuestos(canales)

    def seleccionarTarjeta(self, divisa, retirar):
        """
        Selecciona las tarjetas disponibles para realizar una transacción en una divisa determinada.

        Args:
            divisa (str): Divisa de la transacción.
            retirar (bool): Indica si se trata de una transacción de retiro (True) o no (False).

        Returns:
            list: Lista de tarjetas disponibles para realizar la transacción.
        """
        retorno = []
        for t in self.tarjetasDebito:
            if t.divisa == divisa:
                retorno.append(t)
        if not retirar:
            for t in self.tarjetasCredito:
                if t.divisa == divisa:
                    retorno.append(t)
        return retorno

